### ISSUE TITLE
Transfer fail from S3 to Google Storage - Corrupted on transfer: md5sums differ

### ISSUE THREAD

**Issue by peimurphy** [ncw/rclone#56]
Just checking out the rclone application for the first time today to mirror some AWS S3 bucket to Google Cloud Storage and came across issues with larger files not transferring over with the following error - Corrupted on transfer: md5sums differ

I tried on a couple of source S3 buckets today and had the same issue - small files seemed to transfer OK but larger files kept giving this md5sums differ error.

Is there a configuration option I missed or a new issue?

Thanks
Jim

Sample log:
root@Automation:~/rclone-v1.12-linux-amd64# ./rclone sync AWS-S3:db-backups Google-Storage-Test:testoffsitestorage -v

2015/05/07 12:55:19 Storage bucket testoffsitestorage: Modify window is 1ns
2015/05/07 12:55:19 Storage bucket testoffsitestorage: Building file list
2015/05/07 12:55:19 Storage bucket testoffsitestorage: Waiting for checks to finish
2015/05/07 12:55:19 testdb.sql.gz: Size and modification time the same (differ by 0, within tolerance 1ns)
2015/05/07 12:55:19 testdb.sql.gz: Unchanged skipping
2015/05/07 12:55:19 mysql.sql.gz: Size and modification time the same (differ by 0, within tolerance 1ns)
2015/05/07 12:55:19 mysql.sql.gz: Unchanged skipping
2015/05/07 12:55:19 tmp.sql.gz: Size and modification time the same (differ by 0, within tolerance 1ns)
2015/05/07 12:55:19 tmp.sql.gz: Unchanged skipping
2015/05/07 12:55:20 innodb.sql.gz: Size and modification time the same (differ by 0, within tolerance 1ns)
2015/05/07 12:55:20 innodb.sql.gz: Unchanged skipping
2015/05/07 12:55:20 Storage bucket testoffsitestorage: Waiting for transfers to finish
2015/05/07 12:55:31 mw.sql.gz: Corrupted on transfer: md5sums differ "86b051dddf5ce98625874650d15546d2-56" vs "c5b1804b2e32a7b6955c429533ab40af"
2015/05/07 12:55:31 mw.sql.gz: Removing failed copy
2015/05/07 12:55:31 Storage bucket testoffsitestorage: Not deleting files as there were IO errors

Transferred:    464449892 Bytes (37291.96 kByte/s)
Errors:                 1
Checks:                 4
Transferred:            1
Elapsed time:  12.162522363s

2015/05/07 12:55:31 Go routines at exit 9

List of files from the S3 bucket source
2015-05-07 12:30:03        457 innodb.sql.gz
2015-05-07 12:30:05     719871 mysql.sql.gz
2015-05-07 12:30:07        457 testdb.sql.gz
2015-05-07 12:30:08        456 tmp.sql.gz
2015-05-07 12:31:19  464449892 mw.sql.gz

**Comment by ncw** [comment]
Something strange is going on...  `86b051dddf5ce98625874650d15546d2-56` isn't even a valid MD5SUM.  That was reported by the source which was AWS.

This looks like it describes the problem
  * http://stackoverflow.com/questions/6591047/etag-definition-changed-in-amazon-s3

So I would guess that you uploaded the file using a multipart uploader into S3.  rclone is expecting that it was uploaded in a single part and so has an MD5SUM for an Etag.

I could blank the md5sum if I detect one with a `-` in, which would cause that check to be skipped which is probably what you want.

What do you think?

**Comment by peimurphy** [comment]
For our source files in this case, we use mysqldump locally, gzip the file and use the aws cli tools as follows to upload to S3 in a bash script. That sounds exactly what is happening...

From our cron logs
+ /usr/local/bin/aws s3 cp /mnt/tmp/db/mw.sql.gz s3://db-backups/
Completed 7 of 64 part(s) with 1 file(s) remaining
Completed 13 of 64 part(s) with 1 file(s) remaining
Completed 19 of 64 part(s) with 1 file(s) remaining
Completed 25 of 64 part(s) with 1 file(s) remaining
Completed 31 of 64 part(s) with 1 file(s) remaining
Completed 37 of 64 part(s) with 1 file(s) remaining
Completed 44 of 64 part(s) with 1 file(s) remaining
Completed 50 of 64 part(s) with 1 file(s) remaining
Completed 60 of 64 part(s) with 1 file(s) remaining
upload: ../mnt/tmp/db/mw.sql.gz to s3://db-backups/mw.sql.gz

That seems to be what other applications have been doing to handle the issue.

**Comment by ncw** [comment]
I have replicated this by uploading a file with `s3cmd`

```
s3cmd --multipart-chunk-size-mb=5 put bigfile  s3://testbucket

rclone md5sum s3:testbucket
492cb23b77fdf5abbd91e67239457f26-13  bigfile

rclone -v copy s3:testbucket /tmp/testbucket
...
2015/05/09 10:30:37 bigfileb: Corrupted on transfer: md5sums differ "492cb23b77fdf5abbd91e67239457f26-13" vs "fdfbade8da873cc0ca8c0e3047b2ebc9"
2015/05/09 10:30:37 bigfileb: Removing failed copy

```

**Comment by ncw** [comment]
And after the fix, md5sums are reported as blank

```
rclone md5sum s3:testbucket
                                  bigfile
```

Copying now proceeds just fine

```
rclone -v copy s3:testbucket /tmp/testbucket
...
2015/05/09 10:33:45 bigfile: Invalid md5sum (probably multipart uploaded) - ignoring: "492cb23b77fdf5abbd91e67239457f26-13"
2015/05/09 10:33:45 bigfile: Copied (new)
```

**Comment by ncw** [comment]
I've fixed this in v1.13 - please re-open the ticket if the fix isn't correct.

Thanks

Nick

**Comment by peimurphy** [comment]
Looking good so far - thank you for the quick fix.
